pkg/controller/clusterset/clusterset.go

```diff
@@ -0,0 +1,152 @@
+package clusterset
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rancher/k3k/pkg/apis/k3k.io/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	clusterSetController    = "k3k-clusterset-controller"
+	networkPolicyName       = "k3k-cluster-netpol"
+	allTrafficCIDR          = "0.0.0.0/0"
+	maxConcurrentReconciles = 1
+)
+
+type ClusterSetReconciler struct {
+	Client      ctrlruntimeclient.Client
+	Scheme      *runtime.Scheme
+	ClusterCIDR string
+}
+
+// Add adds a new controller to the manager
+func Add(ctx context.Context, mgr manager.Manager, clusterCIDR string) error {
+	// initialize a new Reconciler
+	reconciler := ClusterSetReconciler{
+		Client:      mgr.GetClient(),
+		Scheme:      mgr.GetScheme(),
+		ClusterCIDR: clusterCIDR,
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.ClusterSet{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: maxConcurrentReconciles,
+		}).
+		Complete(&reconciler)
+}
+
+func (c *ClusterSetReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	klog.Infof("%#v", req)
+	var clusterSet v1alpha1.ClusterSet
+	if err := c.Client.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace}, &clusterSet); err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to get the clusterset: %w", err)
+	}
+
+	klog.Infof("got a clusterset: %v", clusterSet)
+	// create network policy
+	setNetworkPolicy, err := c.netpol(ctx, &clusterSet)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to make a networkpolicy for cluster set: %w", err)
+	}
+	if err := c.Client.Create(ctx, setNetworkPolicy); err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to create networkpolicy for clusterset: %w", err)
+	}
+	if clusterSet.Spec.MaxLimits != nil {
+		quota := v1.ResourceQuota{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "clusterset-quota",
+				Namespace: clusterSet.Namespace,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						UID:        clusterSet.UID,
+						Name:       clusterSet.Name,
+						APIVersion: clusterSet.APIVersion,
+						Kind:       clusterSet.Kind,
+					},
+				},
+			},
+		}
+		quota.Spec.Hard = clusterSet.Spec.MaxLimits
+		if err := c.Client.Create(ctx, &quota); err != nil {
```

Comment:
I think that this needs to handle the case where the quota already exists but we wanted to update the limits (i.e. check for conflict, and on conflict update). Do see the above note though on the name - if we don't fix the name to be exclusive to one clusterset that could cause problems.